Import rust-installer & adjust compression settings

Mark-Simulacrum opened the pull request:
This brings in rust-lang/rust-installer#123, which enables a larger compression window (8 -> 64MB) amongst other changes to the xz compression settings. The net effect should be smaller compressed tarballs which will decrease bandwidth usage for
static.rust-lang.org, download times, and decompression time.

This comes at the cost of higher baseline requirements for running rustup to use these files, which we believe should be largely acceptable (running rustc is likely to use at least this much memory) but if we get specific reports we may explore options to decrease impact (e.g., using the gzip tarballs automatically in rustup).

To simplify iteration on compression settings this also imports the rust-lang/rust-installer submodule, it is now hosted fully inside rust-lang/rust. Once we land this I'll file a followup to add a note to that repo and we can subsequently archive it.

--

CI times for dist-x86_64-linux builds:

* threads=6, master - 2h 50m
* threads=1, new - 3h 40m
* threads=6, new - 2h 50m

Mark-Simulacrum committed:
Bump rust-installer

This brings in rust-lang/rust-installer#123, which enables a larger
compression window (8 -> 64MB) amongst other changes to the xz
compression settings. The net effect should be smaller compressed
tarballs which will decrease bandwidth usage for
static.rust-lang.org, download times, and decompression time.

This comes at the cost of higher baseline requirements for running
rustup to use these files, which we believe should be largely acceptable
(running rustc is likely to use at least this much memory) but if we get
specific reports we may explore options to decrease impact (e.g., using
the gzip tarballs automatically in rustup).

pietroalbini commented:
@bors r+

bors commented:
:pushpin: Commit 48af38fa0c292369f7ce992afe832bb8f023a1f8 has been approved by `pietroalbini`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pietroalbini 48af38fa0c292369f7ce992afe832bb8f023a1f8 -->
<!-- homu: {"type":"Approved","sha":"48af38fa0c292369f7ce992afe832bb8f023a1f8","approver":"pietroalbini","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Mark-Simulacrum commented:
@bors r-

I want to check locally with dev-static

Mark-Simulacrum commented:
@bors try

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4285139862/jobs/7466974368)
Build commit: 9060c7f5f9eae7cc9697bef19898ba1e38b2ee56 (`9060c7f5f9eae7cc9697bef19898ba1e38b2ee56`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4285139862/jobs/7466974368"},"merge_sha":"9060c7f5f9eae7cc9697bef19898ba1e38b2ee56"} -->

Mark-Simulacrum commented:
Baseline install of nightly-2023-02-27 takes 53 MB with `RUSTUP_IO_THREADS=1 RUSTUP_UNPACK_RAM=33554432 rustup update nightly-2023-02-27`. Installing the nightly from the try build with the same environment settings takes 102 MB. So we've increased max rss for a nightly install by around 50 MB. Overall I think this fits within the parameters we were aiming for in infra discussion.

Wall-time for installation seems to take roughly the same amount of time with the minimum settings (6.9s before, 6.4s now). When not constraining I/O threads or memory, 7.2s before, 6.5s now.

Ultimately this all seems fairly reasonable so going to go ahead and move ahead, if we run into trouble we can back it out.

@bors r=pietroalbini

bors commented:
:pushpin: Commit 48af38fa0c292369f7ce992afe832bb8f023a1f8 has been approved by `pietroalbini`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pietroalbini 48af38fa0c292369f7ce992afe832bb8f023a1f8 -->
<!-- homu: {"type":"Approved","sha":"48af38fa0c292369f7ce992afe832bb8f023a1f8","approver":"pietroalbini","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors rollup=never
it looks like this blows up ci times

master:
 [TIMING] dist::PlainSourceTarball PlainSourceTarball -- 123.972
 
 my rollup which included this pr:
 [TIMING] dist::PlainSourceTarball PlainSourceTarball -- 1442.916
 
 this is like more then factor 10 for just this tarball? Just compression takes almost half an hours now.

Mark-Simulacrum commented:
Hm, probably due to dropping the thread count. I'll update to avoid that.

Mark-Simulacrum commented:
@bors r-

Mark-Simulacrum committed:
Directly import rust-installer submodule

This moves the rust-installer code to be directly hosted in
rust-lang/rust, since it's not used elsewhere and this makes it easier
to make and review changes without needing a separate upstream commit.

Mark-Simulacrum committed:
Add bootstrap support for rust-installer

Mark-Simulacrum committed:
Use 3 or 6 compression threads for rust-installer

Limit to 3 threads for 32-bit platforms, otherwise we use 6 threads.
This avoids exceeding the amount of memory we can allocate on a 32-bit
platform.

Mark-Simulacrum commented:
@bors try

bors commented:
:hourglass: Trying commit 1e3a1424cbd0fb977dc3decc8d738b528bd0cd1b with merge 75cfb5a92005373d63c2350e90b444aa0148c563...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1e3a1424cbd0fb977dc3decc8d738b528bd0cd1b","merge_sha":"75cfb5a92005373d63c2350e90b444aa0148c563"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4299380973/jobs/7496391982)
Build commit: 75cfb5a92005373d63c2350e90b444aa0148c563 (`75cfb5a92005373d63c2350e90b444aa0148c563`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4299380973/jobs/7496391982"},"merge_sha":"75cfb5a92005373d63c2350e90b444aa0148c563"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4299368564/jobs/7494503900) [(plain)](https://github.com/rust-lang/rust/commit/1e3a1424cbd0fb977dc3decc8d738b528bd0cd1b/checks/11673563938/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:ac593985615ec2ede58e132d2e21d2b1cbd6127c)
Download action repository 'rust-lang/simpleinfra@master' (SHA:e58f560dd5e72b61cf9aeebf432d862b23ac76a8)
Complete job name: PR (x86_64-gnu-llvm-14, false, ubuntu-20.04-xl)
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: x86_64-gnu-llvm-14
---
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s



failed to execute command: CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" "src/tools/rust-installer/test.sh"
error: No such file or directory (os error 2)

Build completed unsuccessfully in 0:37:03
```

</details>

Mark-Simulacrum commented:
@bors try

bors commented:
:hourglass: Trying commit 7d73d392c2b11cab387acac32c623fd018a14165 with merge e2981ca4ac23abeb9519f967fd60c265124cbea7...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7d73d392c2b11cab387acac32c623fd018a14165","merge_sha":"e2981ca4ac23abeb9519f967fd60c265124cbea7"} -->

klensy commented on src/bootstrap/test.rs:
```diff
@@ -962,7 +962,7 @@ fn compare_browser_ui_test_version(installed_version: &str, src: &Path) {
             if v.trim() != installed_version {
                 eprintln!(
-                    "⚠️ Installed version of browser-ui-test (`{}`) is different than the \
+                    "â ï¸ Installed version of browser-ui-test (`{}`) is different than the \
```
Broken

klensy commented on src/bootstrap/test.rs:
```diff
@@ -2549,7 +2549,7 @@ impl Step for Bootstrap {
         if let Some(flags) = option_env!("RUSTFLAGS") {
             // Use the same rustc flags for testing as for "normal" compilation,
-            // so that Cargo doesn’t recompile the entire dependency graph every time:
+            // so that Cargo doesnât recompile the entire dependency graph every time:
```
and here

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4304437371/jobs/7509699303)
Build commit: e2981ca4ac23abeb9519f967fd60c265124cbea7 (`e2981ca4ac23abeb9519f967fd60c265124cbea7`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4304437371/jobs/7509699303"},"merge_sha":"e2981ca4ac23abeb9519f967fd60c265124cbea7"} -->

klensy commented:
Feels wrong that this thing not broken
https://github.com/rust-lang/rust/blob/5423745db8b434fcde54888b35f518f00cce00e4/src/bootstrap/tool.rs#L425

Mark-Simulacrum commented:
External tool controls things like lints, it should be fine to leave that as is. Need to fix the tests though.

Mark-Simulacrum commented:
@pietroalbini Alright, I think this is good to go now - tests seem to pass and CI times from try build with n=6 threads also match pre-change times.

I'll probably look at some cleanup (rustfmt, lints, maybe inlining the whole code into rustbuild proper) in future PRs but this seems good for now.

klensy commented on src/tools/rust-installer/.gitignore:
```diff
@@ -0,0 +1,5 @@
+tmp
+target/
+**/*.rs.bk
+Cargo.lock
```
Add `rust-installer` to https://github.com/rust-lang/rust/blob/master/Cargo.toml or keep `Cargo.lock`?

Mark-Simulacrum replied:
It's already a member of the top-level workspace. We could adjust the ignore file here to remove these but I don't think it's worth worrying over.

pietroalbini commented:
@bors r+

bors commented:
:pushpin: Commit 287081dbaf814e427d3f8b814babb25d897ddbd3 has been approved by `pietroalbini`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pietroalbini 287081dbaf814e427d3f8b814babb25d897ddbd3 -->
<!-- homu: {"type":"Approved","sha":"287081dbaf814e427d3f8b814babb25d897ddbd3","approver":"pietroalbini","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 287081dbaf814e427d3f8b814babb25d897ddbd3 with merge eaee3b5b5181501a8aa2718e7656e19773525ac2...
<!-- homu: {"type":"BuildStarted","head_sha":"287081dbaf814e427d3f8b814babb25d897ddbd3","merge_sha":"eaee3b5b5181501a8aa2718e7656e19773525ac2"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4339632605/jobs/7577623066)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4339632605/jobs/7577623066","builder_name":"checks-actions"} -->

Mark-Simulacrum commented:
@bors r=pietroalbini rollup=never

bors commented:
:pushpin: Commit 70403375ae66f7e480c0ba616df5f1773895bb02 has been approved by `pietroalbini`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pietroalbini 70403375ae66f7e480c0ba616df5f1773895bb02 -->
<!-- homu: {"type":"Approved","sha":"70403375ae66f7e480c0ba616df5f1773895bb02","approver":"pietroalbini","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`dist-i686-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4339632605/jobs/7577408264) [(plain)](https://github.com/rust-lang-ci/rust/commit/eaee3b5b5181501a8aa2718e7656e19773525ac2/checks/11780279236/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Finished release [optimized] target(s) in 20.67s
[TIMING] tool::ToolBuild { compiler: Compiler { stage: 0, host: i686-unknown-linux-gnu }, target: i686-unknown-linux-gnu, tool: "rust-installer", path: "src/tools/rust-installer", mode: ToolBootstrap, is_optional_tool: false, source_type: Submodule, extra_features: [], allow_features: "" } -- 20.698
[TIMING] tool::RustInstaller { compiler: Compiler { stage: 0, host: i686-unknown-linux-gnu }, target: i686-unknown-linux-gnu } -- 0.000
Dist rust-docs-nightly-i686-unknown-linux-gnu
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Mem', /cargo/registry/src/index.crates.io-1cd66030c949c28d/xz2-0.1.6/src/write.rs:128:18
Build completed unsuccessfully in 0:19:28
```

</details>

Mark-Simulacrum commented:
@bors r=pietroalbini rollup=never

bors commented:
:pushpin: Commit 70403375ae66f7e480c0ba616df5f1773895bb02 has been approved by `pietroalbini`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pietroalbini 70403375ae66f7e480c0ba616df5f1773895bb02 -->
<!-- homu: {"type":"Approved","sha":"70403375ae66f7e480c0ba616df5f1773895bb02","approver":"pietroalbini","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 70403375ae66f7e480c0ba616df5f1773895bb02 with merge e4666f4fe99c67dea739cac0d674f59ce0fe75ca...
<!-- homu: {"type":"BuildStarted","head_sha":"70403375ae66f7e480c0ba616df5f1773895bb02","merge_sha":"e4666f4fe99c67dea739cac0d674f59ce0fe75ca"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4350115925/jobs/7601178730)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4350115925/jobs/7601178730","builder_name":"checks-actions"} -->

Mark-Simulacrum commented:
@bors r=pietroalbini rollup=never

bors commented:
:pushpin: Commit dabafb44e0917709910915dc5b3bc351a519a9e1 has been approved by `pietroalbini`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pietroalbini dabafb44e0917709910915dc5b3bc351a519a9e1 -->
<!-- homu: {"type":"Approved","sha":"dabafb44e0917709910915dc5b3bc351a519a9e1","approver":"pietroalbini","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`i686-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4350115925/jobs/7600505977) [(plain)](https://github.com/rust-lang-ci/rust/commit/e4666f4fe99c67dea739cac0d674f59ce0fe75ca/checks/11810411846/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s



failed to execute command: "D:\\a\\rust\\rust\\src/tools/rust-installer/test.sh"
error: %1 is not a valid Win32 application. (os error 193)

Build completed unsuccessfully in 0:57:32
Build completed unsuccessfully in 0:57:32
make: *** [Makefile:68: ci-subset-1] Error 1
```

</details>

bors commented:
:hourglass: Testing commit dabafb44e0917709910915dc5b3bc351a519a9e1 with merge 60445fda58fc9f9aa413752530d6756ae5570f2f...
<!-- homu: {"type":"BuildStarted","head_sha":"dabafb44e0917709910915dc5b3bc351a519a9e1","merge_sha":"60445fda58fc9f9aa413752530d6756ae5570f2f"} -->

rust-timer commented:
Finished benchmarking commit (60445fda58fc9f9aa413752530d6756ae5570f2f): [comparison URL](https://perf.rust-lang.org/compare.html?start=9b60e6c68ff0aabad9a0edd71898466886dbf6bb&end=60445fda58fc9f9aa413752530d6756ae5570f2f&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=9b60e6c68ff0aabad9a0edd71898466886dbf6bb&end=60445fda58fc9f9aa413752530d6756ae5570f2f&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9b60e6c68ff0aabad9a0edd71898466886dbf6bb&end=60445fda58fc9f9aa413752530d6756ae5570f2f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.3%  | [1.3%, 1.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.2% | [-3.2%, -3.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9b60e6c68ff0aabad9a0edd71898466886dbf6bb&end=60445fda58fc9f9aa413752530d6756ae5570f2f&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

klensy commented:
> Feels wrong that this thing not broken
> 
> https://github.com/rust-lang/rust/blob/5423745db8b434fcde54888b35f518f00cce00e4/src/bootstrap/tool.rs#L425
> 
> https://github.com/rust-lang/rust/blob/5423745db8b434fcde54888b35f518f00cce00e4/src/bootstrap/lib.rs#L486

This comment strike back: while working on #112884 i found out that this flag set `SourceType::Submodule` for rust-installer and this is cause why rustfmt didn't format it's sources. It seems there no working assert that submodule is actually submodule.
